src/ports/postgres/modules/sample/balance_sample.py_in

```diff
@@ -0,0 +1,748 @@
+                            set(desired_level_counts.keys()))
+        if output_table_size:
+            # Uniformly distribute across the remaining class levels
+            remaining_rows = output_table_size - sum(desired_level_counts.values())
+            if remaining_rows > 0:
+                rows_per_level = math.ceil(float(remaining_rows) /
+                                           len(remaining_levels))
+                for each_level in remaining_levels:
+                    sample_strategy = _choose_strategy(
+                        actual_level_counts[each_level], rows_per_level)
+                    target_level_counts[each_level] = (rows_per_level,
+                                                       sample_strategy)
+        else:
+            # When output_table_size is unspecified, rows from the input table
+            # are sampled as is for remaining class levels. This is same as the
+            # NOSAMPLE strategy.
+            for each_level in remaining_levels:
+                target_level_counts[each_level] = (actual_level_counts[each_level],
+                                                    NOSAMPLE)
+    else:
+        def ceil_of_mean(numbers):
+            return math.ceil(float(sum(numbers)) / max(len(numbers), 1))
+
+        # UNIFORM: Ensure all level counts are same (size determined by output_table_size)
+        # UNDERSAMPLE: Ensure all level counts are same as the minimum count
+        # OVERSAMPLE: Ensure all level counts are same as the maximum count
+        size_function = {UNDERSAMPLE: min,
+                         OVERSAMPLE: max,
+                         UNIFORM: ceil_of_mean
+                         }[sampling_strategy_str]
+        if sampling_strategy_str == UNIFORM and output_table_size:
+            # Ignore actual counts for computing target sizes
+            # if output_table_size is specified
+            target_size_per_level = math.ceil(float(output_table_size) /
+                                              len(actual_level_counts))
+        else:
+            target_size_per_level = size_function(actual_level_counts.values())
+        for each_level, actual_count in actual_level_counts.items():
+            sample_strategy = _choose_strategy(actual_count, target_size_per_level)
+            target_level_counts[each_level] = (target_size_per_level,
+                                               sample_strategy)
+    return target_level_counts
+
+# -------------------------------------------------------------------------
+
+
+def _get_sampling_strategy_specific_dict(target_class_sizes):
+    """ Return three dicts, one each for undersampling, oversampling, and
+        nosampling. The dict contains the number of samples to be drawn for
+        each class level.
+    """
+    undersample_level_dict = {}
+    oversample_level_dict = {}
+    nosample_level_dict = {}
+    for level, (count, strategy) in target_class_sizes.items():
```

Comment:
Minor suggestion : an alternative to using the chosen_strategy dict is to assign all the three dicts in the if check itself. something like
```
    for level, (count, strategy) in target_class_sizes.items():
        if strategy == UNDERSAMPLE:
            undersample_level_dict[level] = count
        elif strategy == OVERSAMPLE:
            oversample_level_dict[level] = count
        else:
            nosample_level_dict[level] = count

    return (undersample_level_dict, oversample_level_dict, nosample_level_dict)
```
It is a bit redundant but more direct. I will leave the decision up to you